### PR TITLE
feat: fix incorrect node loc

### LINKED ISSUE #287: warning  `>` can be escaped with `&gt;`
### Subject of the issue

Here's the content of a mdx file:

```mdx
Webpack 5 has been officially released. Read our <a href="/blog/2020-10-10-webpack-5-release/">announcement</a>. Not ready yet? Read <a href="https://v4.webpack.js.org/">webpack 4 documentation here</a>.
```

And the eslint configuration:

```js
module.exports = {
  root: true,
  extends: ['eslint:recommended', 'plugin:react/recommended', 'prettier'],
  parser: '@babel/eslint-parser',
  env: {
    browser: true,
    es6: true,
    node: true,
    jest: true,
    'cypress/globals': true,
  },
  plugins: ['markdown', 'cypress', 'react-hooks'],
  globals: {
    __DEV__: true,
  },
  rules: {
    'no-console': 'off',
    semi: ['error', 'always'],
    quotes: ['error', 'single'],
    'react/jsx-uses-react': 'off', // no longer needed with new jsx transform
    'react/react-in-jsx-scope': 'off', // ditto
    'react-hooks/rules-of-hooks': 'error',
    'react-hooks/exhaustive-deps': 'warn',
  },
  settings: {
    react: {
      version: 'detect',
    },
  },
  overrides: [
    { files: ['src/**/*.jsx'] }, // eslint would lint .js only by default
    {
      files: ['**/*.mdx'],
      extends: ['plugin:mdx/recommended'],
      globals: {
        Badge: true,
      },
      rules: {
        semi: ['off'],
      },
      settings: {
        'mdx/code-blocks': true,
      },
    },
    {
      files: ['**/*.md'],
      processor: 'markdown/markdown',
    },
    {
      files: ['**/*.{md,mdx}/*.{js,javascript}'], // we don't lint ts at the moment
      rules: {
        indent: ['error', 2],
        quotes: ['error', 'single'],
        'no-undef': 'off',
        'no-unused-vars': 'off',
        'no-constant-condition': 'off',
        'no-useless-escape': 'off',
        'no-dupe-keys': 'off',
      },
    },
  ],
};
```
And here's the lint result:

```bash
  1:94   warning  `"` can be escaped with `&quot;`, `&ldquo;`, `&#34;`, `&rdquo;`  mdx/no-unescaped-entities
  1:95   warning  `>` can be escaped with `&gt;`                                   mdx/no-unescaped-entities
  1:111  warning  `>` can be escaped with `&gt;`                                   mdx/no-unescaped-entities
```

### Your environment

*   **OS**: 
*   **Packages**: 
*   **Env**: 

### Steps to reproduce

Here's the pull request https://github.com/webpack/webpack.js.org/pull/4326 you can use to reproduce:

1. clone the pull request
2. run `yarn` to install
3. run `yarn lint:js`

### Expected behaviour

No warning showed.

### Actual behaviour

Warnings.

### PR DESCRIPTION
**Pull request by JounQin**
this causes issue with `react/no-unescaped-entities` before
so now `mdx/no-unescaped-entities` and `eslint-plugin-react` dep can be removed!

fix #287

### CODE REVIEW

**Comment by JounQin** [comment]
@mdx-js/core Can you help to enable https://codesandbox.io/docs/ci for this repo?

**Comment by wooorm** [comment]
Why is CS needed?

**Comment by JounQin** [comment]
> Why is CS needed?

I want to publish to codesanbox's registry for testing for every PR.

**Comment by wooorm** [comment]
What does CodeSandbox check that `npm test` or `yarn test` doesn’t?

**Comment by JounQin** [comment]
> What does CodeSandbox check that `npm test` or `yarn test` doesn’t?

My purpose is not related to `npm test` or `yarn test`, I mean that we can provide a testing version for users to confirm it's working as expected before publishing it to true npm registry for eveny PR.

**Comment by JounQin** [comment]
@wooorm Can we merge and release first because it fixes a bug which breaks ESLint?

**Comment by ChristianMurphy** [comment]
> I want to publish to codesanbox's registry for testing for every PR without publishing to npm.

CodeSandbox can point to a specific branch without an integration using https://codesandbox.io/s/github
https://codesandbox.io/s/github/mdx-js/eslint-mdx/tree/fix/no-unescaped-entities for this branch at the root level, it can also be pointed to specific folders/paths for example https://codesandbox.io/s/github/mdx-js/eslint-mdx/tree/fix/no-unescaped-entities/packages/eslint-plugin-mdx

Is there a specific feature you're looking for from the integration?

**Comment by JounQin** [comment]
> Is there a specific feature you're looking for from the integration?

It seems nope. I want an installable registry instead of editing source codes.

File: packages/eslint-mdx/src/helpers.ts [hunk 1]
```diff
@@ -112,23 +112,34 @@ export const restoreNodeLocation = <T>(
...
     },
   })
 }
 
+export const arrayify = <T, R = T extends Array<infer S> ? S : T>(
+  ...args: T[]
+) =>
+  args.reduce<R[]>((arr, curr) => {
+    arr.push(...(Array.isArray(curr) ? curr : curr == null ? [] : [curr]))
+    return arr
```
Comment (ChristianMurphy) [inline packages/eslint-mdx/src/helpers.ts:140]: I'd avoid nesting terneries if possible, sometimes using multiple returns can be clearer

```suggestion
    if (Array.isArray(curr)) return arr.concat(curr)
    if (curr == null) return arr
    return arr.concat([curr])
```

File: packages/eslint-plugin-mdx/src/rules/helpers.ts [hunk 2]
```diff
@@ -90,9 +91,10 @@ export const getRemarkProcessor = (searchFrom: string, isMdx: boolean) => {
 
   return plugins
     .reduce((processor, pluginWithSettings) => {
-      const [plugin, ...pluginSettings] = Array.isArray(pluginWithSettings)
-        ? pluginWithSettings
-        : [pluginWithSettings]
+      const [plugin, ...pluginSettings] = arrayify(pluginWithSettings) as [
+        RemarkPlugin,
+        ...unknown[]
+      ]
```
Comment (ChristianMurphy) [inline packages/eslint-plugin-mdx/src/rules/helpers.ts:97]: Does this change make sense?
It goes from a single condition, which does not require a case, and is three lines long.
To a multiple condition, which requires a cast, and is four lines long. :thinking:

File: packages/eslint-mdx/src/helpers.ts [hunk 3]
```diff
@@ -112,23 +112,34 @@ export const restoreNodeLocation = <T>(
...
+        line: restoredEndLine,
+        column: endLoc.column + (restoredEndLine === 1 ? offset : 0),
       },
     },
   })
 }
 
+export const arrayify = <T, R = T extends Array<infer S> ? S : T>(
+  ...args: T[]
+) =>
```
Comment (ChristianMurphy) [inline packages/eslint-mdx/src/helpers.ts:137]: Broadly this seems like a custom implementation of https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat ? :thinking:

Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:137]: But also accept non-array to be array.

File: packages/eslint-plugin-mdx/src/rules/helpers.ts (hunk 2 above)
Comment (JounQin) [reply packages/eslint-plugin-mdx/src/rules/helpers.ts:97]: typings will be dropped on compiling, so the final codes would be simpler.

File: packages/eslint-mdx/src/helpers.ts (hunk 1 above)
Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:140]: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-spread.md

File: packages/eslint-mdx/src/helpers.ts (hunk 3 above)
Comment (ChristianMurphy) [reply packages/eslint-mdx/src/helpers.ts:137]: something like:

```ts
export function arrayify<T, R = T extends Array<infer S> ? S : T>(...args: T[]): R {
  return args.flat();
}
```

?

Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:137]: But also drop `null/undefined` values.

And `Array#flat` is not available on Node 10

https://node.green/#ES2019-features-Array-prototype--flat--flatMap--Array-prototype-flat

https://github.com/mdx-js/eslint-mdx/blob/master/packages/eslint-plugin-mdx/package.json#L14

File: packages/eslint-mdx/src/helpers.ts (hunk 1 above)
Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:140]: I think it's OK for such simple statement.

Comment (ChristianMurphy) [reply packages/eslint-mdx/src/helpers.ts:140]: > https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-spread.md

:man_shrugging: okay, then

```suggestion
    if (Array.isArray(curr)) return [...arr, ...curr]
    if (curr == null) return arr
    return [...arr, curr]
```

Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:140]: This solution iterates the original whole arr everytime, I would still prefer `push` version.

File: packages/eslint-mdx/src/helpers.ts (hunk 3 above)
Comment (ChristianMurphy) [reply packages/eslint-mdx/src/helpers.ts:137]: > And Array#flat is not available on Node 10

And node 10 is nearly EOL https://github.com/nodejs/Release#release-schedule :man_shrugging: 

> But also drop null/undefined values.

```ts
export function arrayify<Parameter, Return = Parameter extends Array<infer Item> ? Item : Parameter>(...args: Parameter[]): Return[] {
  return args.flat().filter((value) => value != null);
}
```

File: packages/eslint-mdx/src/helpers.ts (hunk 1 above)
Comment (ChristianMurphy) [reply packages/eslint-mdx/src/helpers.ts:140]: > I think it's OK for such a simple statement.

I'm somewhat surprised it wasn't flagged by the linter, many configs have https://eslint.org/docs/rules/no-nested-ternary enabled.

**Comment by ChristianMurphy** [comment]
> It seems nope. I want an installable registry instead of editing source codes.

:thinking: but codesandbox isn't an install-able registry.
People wouldn't be able to pull down the version and test it on their documents locally.

File: packages/eslint-mdx/src/helpers.ts (hunk 1 above)
Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:140]: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-nested-ternary.md

Comment (ChristianMurphy) [reply packages/eslint-mdx/src/helpers.ts:140]: Okay?
This code doesn't follow the style guide in either https://eslint.org/docs/rules/no-nested-ternary or https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-nested-ternary.md

**Comment by JounQin** [comment]
> > It seems nope. I want an installable registry instead of editing source codes.
> 
> 🤔 but codesandbox isn't an install-able registry.
> People wouldn't be able to pull down the version and test it on their documents locally.

https://codesandbox.io/docs/ci

> This is published to our registry, so you can immediately test the library in CodeSandbox or locally without publishing to npm or elsewhere.

Did I misunderstand?

File: packages/eslint-mdx/src/helpers.ts (hunk 3 above)
Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:137]: > And node 10 is nearly EOL

But that would be a breaking change, I would drop Node 10 support in v2.

File: packages/eslint-mdx/src/helpers.ts (hunk 1 above)
Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:140]: Seems been disabled by `eslint-plugin-prettier` again due to parens. 🤣

And it follows `the nested ternary is only one level`

**Comment by JounQin** [comment]
codesandbox seems simplest to be integrated to me.

**Comment by ChristianMurphy** [comment]
> codesandbox seems simplest to be integrated to me.

Simplest for who?
It sounds like yet another registry for end users to configure.

> NPM_TOKEN

True, the release team is fairly small at the moment and cautious about automation

> GITHUB_TOKEN

this one could/should be easier?
It's tagline is:
> Streamline your workflow
> Use the same GITHUB_TOKEN for all automated package uploads and downloads through Actions.

**Comment by JounQin** [comment]
> Simplest for who?

For myself. 😂

> this one could/should be easier?

I don't know what permissions are required, so codesanbox is the simplest one.

**Comment by JounQin** [comment]
And if someone is going to set up an automation of GitHub packages, it would also be great to have.

**Comment by JounQin** [comment]
Merge first, need to be released @wooorm

**Comment by JounQin** [comment]
> And if someone is going to set up an automation of npm or GitHub packages, it would also be great to have.

@ChristianMurphy 

Just remember that, only master source codes should be published to npm as canary version.

I still think codesandbox is the best choice for PR.

**Comment by wooorm** [comment]
@JounQin Why not use install from github? https://docs.npmjs.com/cli/v7/configuring-npm/package-json#github-urls

---

Released

**Comment by JounQin** [comment]
@wooorm The source code is all written in `.ts`, no `.js` files are included into git.

**Comment by wooorm** [comment]
Ah 😅 That’s another good reason for using JS (+ JSDoc based TS)!

---

> I still think codesandbox is the best choice for PR.

Do you think this will help a lot of people in the case of `eslint-mdx`?
I do not see a lot of codesandbox reproductions: https://github.com/mdx-js/eslint-mdx/issues?q=is%3Aissue+is%3Aclosed.
Say you wrote a PR to fix some problem, then I’m assuming you add a test for that asserts it’s fixed.
Isn’t that enough?

**Comment by JounQin** [comment]
> Ah 😅 That’s another good reason for using JS (+ JSDoc based TS)!

Writing `.js` with `jsdoc` is a bit painful and bad experience to myself.

---

It is not related to *codesandbox reproductions*, it's just a custom npm registry to install locally.

Tests added in PR only confirm the single case while the true project could be more complex, a pre-canary version is a better choice to me.

**Comment by JounQin** [comment]
Of course, if we merge first, and publish canary version for every commit on master, that will be fine too.

**Comment by JounQin** [comment]
@wooorm @ChristianMurphy 

If we use `codesanbox CI`, we can test PR version locally like `yarn add https://pkg.csb.dev/babel/babel/commit/c6bcf5c9/@babel/cli`.

See https://ci.codesandbox.io/status/babel/babel/pr/13046/builds/112443 for instance.

So, can we enable it for this repo? (I don't understand why it should not.)

**Comment by wooorm** [comment]
Looks cool!

> I don't understand why it should not

Security.

You’re also asking about CodeSandbox, but you don’t care about that specifically, you want to release faster.

- Downside with CSB is that it adds an app to the whole organization which is only used in one repo
- Another one is that PRs are unsafe: anyone can make a PR. It’s dangerous.
- Downside of npm tokens is that they’re per person: if I add a token here, **all** of my packages can be published. If that token is leaked, that’s incredibly bad

Why not publish to the GH package registry? In GH actions, you get access to a token specifically for that repository. They have security in place so that PRs are made safe.

- https://docs.github.com/en/packages/guides/configuring-npm-for-use-with-github-packages#publishing-a-package
- https://stackoverflow.com/questions/46808515/how-to-have-npm-dependencies-from-different-registries-in-the-package-json/65068985#65068985

**Comment by JounQin** [comment]
> Downside with CSB is that it adds an app to the whole organization which is only used in one repo

Is that true? I'm just asking for installing this app on this repo. And it can always be configured later.

> Another one is that PRs are unsafe: anyone can make a PR. It’s dangerous.

That's why I don't want to publish PR versions to npm nor GH registry. The user must install a specific PR version, but won't get updates automatically. We don't need to change registry config at all.

> Downside of npm tokens is that they’re per person: if I add a token here, all of my packages can be published. If that token is leaked, that’s incredibly bad

no npm token is required here

**Comment by JounQin** [comment]
image

**Comment by wooorm** [comment]
>  I'm just asking for installing this app on this repo.

Ah, yes, you’re right. I can add it to one or more specific repos.

> No npm token is required here

I know! But this is to reiterate my stance on npm tokens

---

You did say “Of course, if we merge first, and publish canary version for every commit on master, that will be fine too.” before. That’s different from what you’re saying now. And such a canary version can be done with GH.

So, do you prefer CSB to publish PRs, or GH for canary releases?

**Comment by JounQin** [comment]
> You did say “Of course, if we merge first, and publish canary version for every commit on master, that will be fine too.” before. That’s different from what you’re saying now. And such a canary version can be done with GH.

I say that because it seems you disagree to install CSB into this repo, so I proposed to release canary versions for every commit on master. And it seems true for `mdx`: https://github.com/mdx-js/mdx/blob/main/.github/workflows/publish.yml#L33

> So, do you prefer CSB to publish PRs, or GH for canary releases?

I'd prefer CSB always.

**Comment by JounQin** [comment]
Besides, [Package Scoped tokens](https://github.com/npm/roadmap/issues/10) is now on npm's roadmap.

**Comment by wooorm** [comment]
> Besides, Package Scoped tokens is now on npm's roadmap.

It’s been for half a year, and it’s not moving fast.

---

Added CSB. Here’s an example config `.codesandbox/ci.json` they suggest for monorepos:

```json
{
  // If you have a monorepo we infer your packages from your Yarn workspaces
  // or lerna configuration by default. If you want to explicitly
  // set what to build, you can fill the 'packages' field with paths to your
  // packages
  "packages": ["packages/react", "packages/react-dom"],
  "sandboxes": ["new", "vanilla"]
}
```

**Comment by JounQin** [comment]
> It’s been for half a year, and it’s not moving fast.

Yes, sadly.

---

Thanks for adding CSB! ❤️ 

`.codesandbox/ci.json` has been added in this PR, it's just not enabled.

https://github.com/mdx-js/eslint-mdx/pull/288/files#diff-bca34ff1efba1992c546adca0b1596996778b69817a7c391d29f40806694fdc9

**Comment by wooorm** [comment]
you probably have to open a new PR

**Comment by JounQin** [comment]
Yes, https://github.com/mdx-js/eslint-mdx/pull/299

And it works https://ci.codesandbox.io/status/mdx-js/eslint-mdx/pr/299/builds/112534

Cheers!

**Pull request merged**